src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs

```diff
@@ -131,6 +175,23 @@ public static SafeFreeCredentials AcquireCredentialsHandle(X509Certificate certi
             return AcquireCredentialsHandle(direction, secureCredential);
         }
 
+        internal static byte[] GetNegotiatedApplicationProtocol(SafeDeleteContext context)
+        {
+            Interop.SecPkgContext_ApplicationProtocol alpnContext = SSPIWrapper.QueryContextAttributes(
+                GlobalSSPI.SSPISecureChannel,
+                context,
+                Interop.SspiCli.ContextAttribute.SECPKG_ATTR_APPLICATION_PROTOCOL) as Interop.SecPkgContext_ApplicationProtocol;
+
+            if (alpnContext == null ||
+                alpnContext.ProtoNegoExt != Interop.ApplicationProtocolNegotiationExt.ALPN ||
+                alpnContext.ProtoNegoStatus != Interop.ApplicationProtocolNegotiationStatus.Success)
+            {
+                return null;
```

Comment:
What is the situation where we expect ProtoNegoExt to not be ALPN, or ProtoNegoStatus to not be success?  And should those cases be exceptions?  (Comment seems to have been lost due to formatting fixes)

Reply:
Will add comment, NPN is an earlier version of alpn. Here, the code is querying the context for information about the alpn protocol negotiated, and the win32 api should return object with structure SecPkgContext_ApplicationProtocol, this is just checking if the object returned is indeed of this type and the context of the alpn operation was successful and of type alpn. If there was some internal error that resulted in a wrong object or object with incorrect context returned we'll just return null.